wrapper.go

```diff
@@ -6,58 +6,70 @@ import (
 	"github.com/signalfx/golib/datapoint"
+	log "github.com/sirupsen/logrus"
 	"os"
 	"strings"
 	"time"
 )
 
-// HandlerWrapper is a lambda.Handler implementation that delegates to the embedded lambda.Handler.
-type HandlerWrapper struct {
+// HandlerWrapper extends interface lambda.Handler to support sending metric datapoints.
+type HandlerWrapper interface {
+	Invoke(ctx context.Context, payload []byte) ([]byte, error)
+	SendDatapoints(dps []*datapoint.Datapoint) error
+}
+
+// handlerWrapper is a HandlerWrapper and lambda.Handler implementation.
+// handlerWrapper delegates lambda handler function invocation to the embedded lambda.Handler.
+type handlerWrapper struct {
 	lambda.Handler
 	notColdStart bool
+	ctx          *context.Context
 }
 
-// Invoke is HandlerWrapper's lambda.Handler implementation that delegates to the Invoke method of the embedded lambda.Handler.
+// NewHandlerWrapper is a HandlerWrapper creating factory function.
+func NewHandlerWrapper(handler lambda.Handler) HandlerWrapper {
+	return &handlerWrapper{Handler: handler}
+}
+
+// Invoke is handlerWrapper's lambda.Handler implementation that delegates to the Invoke method of the embedded lambda.Handler.
 // Invoke creates and sends metrics.
-func (hw *HandlerWrapper) Invoke(ctx context.Context, payload []byte) ([]byte, error) {
-	if ctx == nil {
-		return nil, fmt.Errorf("invalid argument. context is nil")
-	}
+func (hw *handlerWrapper) Invoke(ctx context.Context, payload []byte) ([]byte, error) {
+	hw.ctx = &ctx
 	dps := []*datapoint.Datapoint{hw.invocationsDatapoint()}
+	if !hw.notColdStart {
+		dps = append(dps, hw.coldStartsDatapoint())
+		hw.notColdStart = true
+	}
 	start := time.Now()
 	responseBytes, err := hw.Handler.Invoke(ctx, payload)
 	dps = append(dps, hw.durationDatapoint(time.Since(start)))
 	if err != nil {
 		dps = append(dps, hw.errorsDatapoint())
 	}
-	if !hw.notColdStart {
-		dps = append(dps, hw.coldStartsDatapoint())
-		hw.notColdStart = true
-	}
-	if err2 := hw.SendDatapoints(ctx, dps); err2 != nil {
-		var errString string
-		if err != nil {
-			errString = err.Error()
-		}
-		err = fmt.Errorf(strings.Join([]string{errString, err2.Error()}, "\n"))
+	if err2 := hw.sendDatapoints(ctx, dps); err2 != nil {
+		log.Error(err2)
 	}
 	return responseBytes, err
 }
 
 type dimensions map[string]string
 
-// Start takes a handler function and creates a HandlerWrapper which is a lambda.Handler implementation.
-// Start then passes the HandlerWrapper to method lambda.StartHandler
-func Start(handler interface{}) {
-	lambda.StartHandler(&HandlerWrapper{Handler: lambda.NewHandler(handler)})
+// Start takes HandlerWrapper, a lambda.Handler implementation and passes it function lambda.StartHandler
+func Start(handler HandlerWrapper) {
+	lambda.StartHandler(handler)
+}
+
+// SendDatapoints sends custom metric datapoints to SignalFx.
+func (hw *handlerWrapper) SendDatapoints(dps []*datapoint.Datapoint) error {
+	return hw.sendDatapoints(*hw.ctx, dps)
```

Comment:
Get rid of the indirection here and just make this function be what `sendDatapoints` is.

Reply:
`sendDatapoints` already has access to the context via `hw.ctx`, and also context is meant to be passed around, not stuck on a struct usually.  Just pass context where it's needed and don't store it on `hw`.